File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        Map<Character, String> map = new HashMap<>();
        map.put('2', "abc");
        map.put('3', "def");
        map.put('4', "ghi");
        map.put('5', "jkl");
        map.put('6', "mno");
        map.put('7', "pqrs");
        map.put('8', "tuv");
        map.put('9', "wxyz");
        
        if(digits.equals("")){return List.of();}
        Queue<String> list = new LinkedList<>();
        build(map, list, digits.charAt(0));
        for(int i = 1; i < digits.length() ; i++){
            var c = digits.charAt(i);
            var len = list.size();
            for(int j = 0; j < len; j++){
                var str = list.poll();
                list.remove(str);
                for(var letter : map.get(c).toCharArray()){
                    var nstr=str+(String.valueOf(letter));
                    list.add(nstr);
                }
            }
        }
        return (List<String>)list;
    }
    private void build(Map<Character, String> map, Queue<String> list, char digit){
        for(var letter : map.get(digit).toCharArray()){
            list.add(String.valueOf(letter));
        }
    }
}

File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        
        if not digits:
            return []
        
        keyboard = {}
        keyboard['2'] ="abc"
        keyboard['3'] ="def"
        keyboard['4'] ="ghi"
        keyboard['5'] ="jkl"
        keyboard['6'] ="mno"
        keyboard['7'] ="pqrs"
        keyboard['8'] ="tuv"
        keyboard['9'] ="wxyz"
        
        def backtracking(comb : str, indx : int):
            if len(comb) == n:
                ans.append(comb)
                return
            if indx > len(digits):
                return
            
            for word in keyboard[digits[indx]]:
                for letter in word:
                    comb+=letter
                    backtracking(comb, indx+1)
                    comb = comb[:-1]
                
            
        
        n = len(digits)
        ans = []
        backtracking("", 0)
        return ans